### ISSUE TITLE
Bug in Link causing server navigation when app directory + pages present

### ISSUE THREAD

**Issue by myxoh** [vercel/next.js#54231]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Tested on Macbook and Windows, on chrome, safari and edge.

Produced a minimum viable example and tested on next 13.2 to canary (13.18+)
```

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/superwave-team/next-bug

### To Reproduce

- run npm dev or pnpm dev
- Go to http://localhost:3000/channel/test-2
- Navigate to any of the "not reload" links (You see the same ID and no new page load)
- Go to the "reload" link (You see a new ID and the page loading from 0) (On the network tab you see that we are doing a page navigation and loading a new document)

### Describe the Bug

When the app directory and pages app are present, there seems to be some cases where paths that shouldn't overlap are causing conflicts. In our case we detected this (in a minimally reproduceable example) on `/channel/test-2` (Pages) and `/community/new` (App)

When a user tries to load the page in pages, the <Link> component is triggering a full page navigation as opposed to correctly detecting the page should load client side.

### Expected Behavior

All these links should be client side navigations. There should be no conflicts.

**Comment by yjoer** [comment]
I reported a similar one last month. Linking it here for visibility https://github.com/vercel/next.js/issues/53067

**Comment by myxoh** [comment]
You are right @yjoer I searched and somehow missed this one. Should I close mine ?
I'm not sure which one to leave given neither of them have any traction yet 😞

**Comment by yjoer** [comment]
I guess we can leave both open. In my case, the paths causing the problem in my local environment and CodeSandbox were different. It might be possible for this to go away in certain paths and appears in some other paths. 😵

**Comment by myxoh** [comment]
Ok seems this is down to Bloom Filter collisions - at this stage this starts going a bit over my head by @ijjk maybe you have some insights?
